src/OrbitCaptureClient/ApiEventProcessor.cpp

```diff
@@ -0,0 +1,128 @@
+  api_event.encoded_event.args[3] = grpc_api_event.r3();
+  api_event.encoded_event.args[4] = grpc_api_event.r4();
+  api_event.encoded_event.args[5] = grpc_api_event.r5();
+  ProcessApiEvent(api_event);
+}
+
+void ApiEventProcessor::ProcessApiEvent(const orbit_api::ApiEvent& api_event) {
+  orbit_api::EventType event_type = api_event.Type();
+
+  switch (event_type) {
+    case orbit_api::kScopeStart:
+      ProcessStartEvent(api_event);
+      break;
+    case orbit_api::kScopeStop:
+      ProcessStopEvent(api_event);
+      break;
+    case orbit_api::kScopeStartAsync:
+      ProcessAsyncStartEvent(api_event);
+      break;
+    case orbit_api::kScopeStopAsync:
+      ProcessAsyncStopEvent(api_event);
+      break;
+    case orbit_api::kTrackInt:
+    case orbit_api::kTrackInt64:
+    case orbit_api::kTrackUint:
+    case orbit_api::kTrackUint64:
+    case orbit_api::kTrackFloat:
+    case orbit_api::kTrackDouble:
+    case orbit_api::kString:
+      ProcessTrackingEvent(api_event);
+      break;
+    case orbit_api::kNone:
+      ERROR("Received event of type orbit_api::kNone");
+      break;
+    default:
+      ERROR("ApiEvent::EVENT_NOT_SET read from Capture's gRPC stream");
+  }
+}
+
+void ApiEventProcessor::ProcessStartEvent(const orbit_api::ApiEvent& api_event) {
+  synchronous_event_stack_by_tid_[api_event.tid].emplace_back(api_event);
+}
+
+void ApiEventProcessor::ProcessStopEvent(const orbit_api::ApiEvent& stop_event) {
+  std::vector<orbit_api::ApiEvent>& event_stack = synchronous_event_stack_by_tid_[stop_event.tid];
+  if (event_stack.empty()) {
+    // We received a stop event with no matching start event, which is possible if the capture was
+    // started between the event's start and stop times.
+    return;
+  }
+
+  const orbit_api::ApiEvent& start_event = event_stack.back();
+  TimerInfo timer_info = TimerInfoFromEncodedEvent(
+      start_event.encoded_event, start_event.timestamp_ns, stop_event.timestamp_ns, stop_event.pid,
+      stop_event.tid, /*depth*/ event_stack.size() - 1);
+  capture_listener_->OnTimer(timer_info);
+  event_stack.pop_back();
+}
+
+void ApiEventProcessor::ProcessAsyncStartEvent(const orbit_api::ApiEvent& start_event) {
+  const uint64_t event_id = start_event.encoded_event.event.data;
+  asynchronous_events_by_id_[event_id] = start_event;
+}
+
+void ApiEventProcessor::ProcessAsyncStopEvent(const orbit_api::ApiEvent& stop_event) {
+  const uint64_t event_id = stop_event.encoded_event.event.data;
+  if (asynchronous_events_by_id_.count(event_id) == 0) {
+    // We received a stop event with no matching start event, which is possible if the capture was
+    // started between the event's start and stop times.
+    return;
+  }
+
+  orbit_api::ApiEvent& start_event = asynchronous_events_by_id_[event_id];
+  TimerInfo timer_info = TimerInfoFromEncodedEvent(
+      start_event.encoded_event, start_event.timestamp_ns, stop_event.timestamp_ns, stop_event.pid,
+      stop_event.tid, /*depth*/ 0);
```

Comment:
Is the depth really meant to be always zero? Can't the same id be used to nest async scopes?

Reply:
Yes this is meant to be zero. Async scopes are not nested, they end up being stacked by the UI but there is no hierarchical relationships between async events.